src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -16,21 +16,47 @@ namespace AppInstaller::CLI::Workflow
     namespace
     {
         // Creates a Pin appropriate for the context based on the arguments provided
-        Pinning::Pin CreatePin(Execution::Context& context, std::string_view packageId, std::string_view sourceId)
+        Pinning::Pin CreatePin(Execution::Context& context, const Pinning::PinKey& pinKey)
         {
             if (context.Args.Contains(Execution::Args::Type::GatedVersion))
             {
-                return Pinning::Pin::CreateGatingPin({ packageId, sourceId }, context.Args.GetArg(Execution::Args::Type::GatedVersion));
+                return Pinning::Pin::CreateGatingPin(pinKey, context.Args.GetArg(Execution::Args::Type::GatedVersion));
             }
             else if (context.Args.Contains(Execution::Args::Type::BlockingPin))
             {
-                return Pinning::Pin::CreateBlockingPin({ packageId, sourceId });
+                return Pinning::Pin::CreateBlockingPin(pinKey);
             }
             else
             {
-                return Pinning::Pin::CreatePinningPin({ packageId, sourceId });
+                return Pinning::Pin::CreatePinningPin(pinKey);
             }
         }
+
+        // Gets a search request that can be used to find the installed package
+        // that corresponds with a pin. The search is based on the extra ID
+        // (Product Code or Package Family Name) if it is available; otherwise
+        // it uses the package ID. Using the extra ID allows us to disambiguate
+        // in certain cases.
+        SearchRequest GetSearchRequestForPin(const Pinning::PinKey& pinKey)
+        {
+            SearchRequest searchRequest;
+
+            switch (pinKey.ExtraIdType)
+            {
+            case Pinning::ExtraIdStringType::ProductCode:
+                searchRequest.Filters.emplace_back(PackageMatchField::ProductCode, MatchType::CaseInsensitive, pinKey.ExtraId);
+                break;
+            case Pinning::ExtraIdStringType::PackageFamilyName:
+                searchRequest.Filters.emplace_back(PackageMatchField::PackageFamilyName, MatchType::CaseInsensitive, pinKey.ExtraId);
+                break;
+            case Pinning::ExtraIdStringType::None:
+            default:
+                searchRequest.Filters.emplace_back(PackageMatchField::Id, MatchType::CaseInsensitive, pinKey.PackageId);

```

Comment:
Is there a reason the ID isn't always added as a filter? I'm thinking of the scenario where an organization has their own source which a user has added. The same application may be available from different sources, with different ID's, a user may want updates from one source but pin the ID from another source. On the other hand, that would require them to add two separate pins - one for each ID.

Unless you think that this case is already covered by `winget upgrade <query> -s <source> --include-pinned`

Reply:
> Is there a reason the ID isn't always added as a filter?

That's what I did at first but I had some issues with it for the WindowsSDK. IIRC, when I used the ID, it would match the available package, which we would then try to correlate to the installed packages, find multiple and not know which one to pick. But using only the product code, we can first match the installed package, then correlate it back to the available package without conflicts.

And for the case you mention, I think it can be handled with `winget pin add <query> -s <source>`

Reply:
Makes sense!